Add `@expo/log-box` canary

krystofwoldrich opened the pull request:
<img width="1512" height="982" alt="Screenshot 2025-10-09 at 15 17 49" src="https://github.com/user-attachments/assets/8c35101e-2411-4c95-8d83-c24046d371a1" />

# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

`@expo/log-box` replaces default React Native RedBox and LogBox implementations with a new unified UI. The goal is to provide readable formatted error message whenever possible.  

# How

<!--
How did you build this feature or fix this bug and why?
-->

All UI and parsing related code is situated in `@expo/log-box`.

### Public interface:

- `@expo/log-box` -> Default library export provides TS source of `useLogBox` hook. (To be used in JS entry files.)
- `@expo/log-box/utils` -> Provides CJS utilities (types, errors to log box data parsing) for `@expo/cli` and `expo/async-require/hmr`.
- `@expo/log-box/swap-rn-logbox` -> Provides RN LogBoxReplacement for the RN JS runtime. (replaced by Expo Metro resolver)

### Internal structure
- `/build` -> CJS for use in `@expo/cli/utils`
- `/dist` -> Prebuilt DOM Component for Android and iOS (ExpoLogBox.bundle)
- `/android` -> Swap of Android RedBox by Expo LogBox DOM Component (used in `ExpoReactHostFactory.kt`)
- `/ios` -> Swap of iOS RedBox by Expo LogBox DOM Component (swizzles RCTRedBox)

How does it work?

Native (iOS and Android)

RedBox
We intercept the `showError*` methods and render the LogBox DOM Component. Both Android and iOS have a native implementation providing the DOM Components APIs (like `webview-wrapper.tsx` in JS).

LogBox DOM Component is prebuilt and ships as `ExpoLogBox.bundle` in `@expo/log-box`. The bundle is packed as asset in debug builds of the native application.

LogBox
During resolve we replace `react-native/Libraries/LogBox/LogBoxInspectorContainer.js` by `@expo/log-box/swap-rn-logbox.js` which presents the bundled prebuilt DOM Component from `@expo/log-box`.

Native release builds
The `ExpoLogBox.bundle` is only packed for Debug builds.

Web
Debug builds include building the DOM Component same as before this PR. But the overlay and the toast are rendered in Shadow DOM to isolate the logbox's css.  

Web release builds
The LogBox DOM Component is not included.

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

Web applications have the new LogBox by default (no flag to disable, it directly replaces our previous web implementation)

Native application have the default RN LogBox by default and the new `@expo/log-box` implementation can be enabled by `EXPO_UNSTABLE_LOG_BOX=1`.

### Expo Log Box controll envs
- `EXPO_UNSTABLE_LOG_BOX` (boolish) -> enables the new UI for native builds (includes packing assets, native RedBox replacement and JS LogBox replacement)
- `EXPO_DEBUG_LOG_BOX` (boolish) -> enables DOM Component inspect (both on iOS and Android RedBox replacements, LogBox replacement is always inspectable at the moment)
- `EXPO_DEVELOP_LOG_BOX` (boolish) -> changes the native implementation to fetch DOM Component from a local dev server (separate from the app dev server, needs to be configured separately)

### To try out the new LogBox

Run this standalone app from the canary release https://github.com/krystofwoldrich/expo-log-box-test-app

Or router-e2e/05-errors https://github.com/expo/expo/blob/f2d1b4abc31ba4f34ef1c363ea3ca7e3112e5d29/apps/router-e2e/__e2e__/05-errors/app/index.tsx 

## Notes (Known issues)


### New Expo Log Box
- ~At the moment to enable the new UI on native, you need to install `react-native-webview`~
  - Changed to use `@expo/dom-webview` [`81905b3` (#39958)](https://github.com/expo/expo/pull/39958/commits/81905b3175f59bb8c7586dcc29fbf89b12271994#diff-c5f79115fa157cc45f11f2d31051d5a7bac264989f9f619125c067a77458a8c1)
- ~StackTrace paths are rendered absolute on first render, when cycling thru the issue relative path is shown~
  - Fixed in [`9bb567f` (#39958)](https://github.com/expo/expo/pull/39958/commits/9bb567f9b938f795bab5272359a53062754e0f09)
- ~Slide down animation seems to be cut out in the standalone app.~
  -- Fixed in [`197f168` (#39958)](https://github.com/expo/expo/pull/39958/commits/197f168539dbdbf068654715c2998c970f98583c)
- L: ExpoLogBox overwrites Dev Client error screen (which is wanted as it provides more formatting for error, but is missing Go to Home button)
- ~H: Errors on mobile sent over the HMR are not parsed correctly the code frame with ANSI chars is in the message~
  - Fixed in [`13962a8` (#39958)](https://github.com/expo/expo/pull/39958/commits/13962a83231353c95a1a8af1a65d2c959e1a1d9b)
- M: HTML errors from Android show the HTML tags as a text, we should extract the text value only. 

### Unrelated to the new impl
- After Reload Android LogBox doesn't show (not related to the new impl, also reproduced with the original LogBox)
- In DOM Components both the old and new LogBox shows in the component's webview.
- Symbolicated stacktraces in Safari and Firefox are off (Chrome works)
- Dev Client Bundling Error Screen doesn't render
- Component Stack symbolication is wrong (off multiple lines) both on iOS and Android (web is correct)
- On Android the LogBox is closed when the app goes to background.  
- RN removes `collapsed` from symbolicated frames
  - https://github.com/facebook/react-native/blob/93ac7db54ead26da002abebb8f987f58b6be3a1d/packages/react-native/Libraries/LogBox/Data/LogBoxLog.js#L51

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

lukmccall reviewed (commented):
Looks good!

EvanBacon reviewed (commented):
Maybe we skip checking in the the `build` directory in git since most users will have the project precompiled and embedded. Not sure if this is also true on web though (?).

krystofwoldrich commented:
> Maybe we skip checking in the the build directory in git since most users will have the project precompiled and embedded. 

`dist` holds the prebuilt DOM Component. `build` has CJS for use in Expo CLI.

> Not sure if this is also true on web though (?).

Correct, web doesn't use the prebuilt.

Since `build` exists only for Expo CLI I minimized the number of file by only exporting the needed `@expo/log-box/utils`.
[`9a6db1c` (#39958)](https://github.com/expo/expo/pull/39958/commits/9a6db1cdbec3488d97ddcd3d4fcd12318413edaa)

krystofwoldrich commented on packages/expo/src/dom/webview-wrapper.tsx:
```diff
@@ -66,7 +69,7 @@ const RawWebView = React.forwardRef<object, Props>((props, ref) => {
   const webviewRef = React.useRef<WebViewRef>(null);
   const domImperativeHandlePropsRef = React.useRef<string[]>([]);
-  const source = { uri: `${getBaseURL()}/${filePath}` };
+  const source = sourceOverride ?? { uri: `${getBaseURL()}/${filePath}` };
```
- https://github.com/expo/expo/pull/40397

krystofwoldrich commented on packages/@expo/log-box/src/ContextPlatform.tsx:
```diff
@@ -0,0 +1,40 @@
+  }, [platform]);
+
+  return (
+    <RuntimePlatformContext.Provider value={{ platform, isNative }}>
```
Will be replaced by 
- https://github.com/expo/expo/pull/40382

Note that we need to supply the EXPO_DOM_OS also from from the native WebViewWrappers.

EvanBacon commented on packages/@expo/log-box/src/logbox-rn-polyfill.tsx:
```diff
@@ -0,0 +1,200 @@
+          platform={process.env.EXPO_OS}
+          devServerUrl={getBaseUrl()}
+          dom={{
+            useExpoDOMWebView: true,
```
Is this better than standard webview?

krystofwoldrich replied:
I opted for the expo dom webview since it's smaller, and this will be pulled into every expo app.

EvanBacon commented on packages/@expo/log-box/src/logbox-rn-polyfill.tsx:
```diff
@@ -0,0 +1,200 @@
+            }, 100);
+          }}
+          onCopyText={(text: string) => {
+            Clipboard.setString(text);
```
```suggestion
            Clipboard.setString(text);
```
This is deprecated now, but probably fine for now.

krystofwoldrich commented on packages/@expo/log-box/android/src/main/expo/modules/logbox/ExpoLogBoxSurfaceDelegate.kt:
```diff
@@ -0,0 +1,318 @@
+          val gson = Gson()
+          val jsonObject = gson.toJson(initialProps)
+
+          val script = """
```
[`dd75190` (#39958)](https://github.com/expo/expo/pull/39958/commits/dd75190a491ceeb0ee02c7e7db3b38d183f2b1df)

krystofwoldrich commented on packages/@expo/log-box/src/logbox-dom-polyfill.tsx:
```diff
@@ -0,0 +1,248 @@
+    };
+  }
+  // @ts-ignore
+  globalThis.__polyfill_dom_reloadRuntime = reloadRuntime;
```
Updated in [`65758d2` (#39958)](https://github.com/expo/expo/pull/39958/commits/65758d29867af661237733c38eb8e5213853688d)

krystofwoldrich commented on packages/@expo/log-box/src/LogBox.ts:
```diff
@@ -0,0 +1,120 @@
+  // Interpolate the message so they are formatted for adb and other CLIs.
+  // This is different than the message.content above because it includes component stacks.
+  // const interpolated = parseInterpolation(args);
+  // originalConsoleError?.(interpolated.message.content);
```
At the moment this is only used on web, so we don't have to interpolate the message.

krystofwoldrich commented on packages/expo/src/async-require/hmr.ts:
- HMR Changes will be removed after https://github.com/expo/expo/pull/40449 is merged.

krystofwoldrich commented on packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts:
- After https://github.com/expo/expo/pull/40449/files#diff-1413d30a2bc0175eecab999051ac255e61477632287b90f2c913e74920322103 this can be replaced by the helper `doReplace`.

krystofwoldrich committed:
Add `@expo/log-box`

last fix tests

Revert "[bare-expo][Android] Turn on build caching (#40206)"

This reverts commit 7a49dc578e525e063225ede06299403268dad6d2.

exclude swift code when new logbox is disabled

exclude swift code when new logbox is disabled

revert use -> useContext

skip the hanging test

resolve expo/log-box safely

use metro resolve

use expo dom webview

fix utils export

fix example app number

fix formatted paths not rendering

use serverRoot instead of projectRoot

inject expo logbox when package is installed on android

enable new logbox when the package is installed

fix kt lint

Revert "enable new logbox when the package is installed"

This reverts commit 74b035586a38f11fd4ea94ac284557dbf9eb99e9.

keep a flag to enable due to web support

apply android review suggestions

fix build

refactor android webview wrapper

fix multiple react version installed

revert expo podspec empty line

ensure setupLogBox is removed for production builds

remove expo log box bundle from git

minimize the build files

explain App.tsx

clean up app.json

Add sync comment to webview wrappers

remove unused imports

use EXPO_OS instead of platform

fix check package

update README and package.json description

remove unnecessary react native config

add script descriptions for build and watch bundle

fix react 19.1, misc refactors

remove install middleware

refactor errors parsing, add comments, split into multiple files

Fix missing component stack in web fatal errors

fix dismiss index manipulation

fix ios slide down animation

Add native hmr fork

fix log box build

remove unused logboxnative

clean up overlay actions props

fix rebase

fix rebase

fix rebase

fix rebase

clean up interpolation comment

enable react compiler

Reapply "[bare-expo][Android] Turn on build caching (#40206)"

This reverts commit 2725b24032c0fd65fe6d5835685e387ce3c40e7b.

use global.css with lowercase g

use expo parser for errors on native (ios, android)

fix parse hmr errors name

fix typo

remove duplicate regexes and metro build error experimental class

simplify appending components stack

do not show code frame for irrelevant stacks

fix stack trace frames list order and collapsing

fix stack trace frames list order and collapsing

fix lock

add EXPO_DOM_HOST_OS to native webview wrappers

remove EXPO_DOM_HOST_OS polyfill

clean up unused files

krystofwoldrich committed:
use existing doReplace helpers

EvanBacon reviewed (approved):
Nice work, exciting. Should circle back in the future and clean up the CSS/styles, but no reason to block this PR.

krystofwoldrich committed:
fix test linked deps

krystofwoldrich committed:
ensure build script works on windows

krystofwoldrich committed:
fix playwright test with spaces in path

krystofwoldrich committed:
bump to latest canary
